Generalize beam search, improve simple_seq2seq

epwalsh opened the pull request:
Updates to `simple_seq2seq`:
- supports attention module
- utilizes beam search for decoding

~~@brendan-ai2 I didn't want to end up re-implementing beam search again, so I used the existing system by defining a simple state class, and then just implementing the `take_step` method within `SimpleSeq2Seq`. Let me what you think!~~

~~@DeNeutoy the test `allennlp/tests/predictors/simple_seq2seq_test.py` is failing because it depends on an old archived model fixture at `allennlp/tests/fixtures/encoder_decoder/simple_seq2seq/serialization`. I'm not sure if you needed that exact model for something else, or if I could just replace it with another trained model.~~

**TODO:**
- [x] pull out and generalize beam search from `event2mind`
- [x] replace `state_machine.beam_search` in `simple_seq2seq` with more efficient version
- [x] update model that predictor test depends on
- [x] add unit tests for new beam search

brendan-ai2 commented:
Hey @epwalsh, sorry I think I may have explained things poorly in https://github.com/allenai/allennlp/issues/1833. The beam search coded up in the `state_machines` package is not intended for simple seq2seq decoding.
To quote the file I linked directly, "Typical seq2seq decoding, where you have a fixed vocabulary and no constraints on your output, can be done much more efficiently than we do in this code." That was the intention behind the model I linked you earlier. See @matt-gardner's comment here: https://github.com/allenai/allennlp/pull/1679#issuecomment-419249544

epwalsh commented:
@brendan-ai2 gotcha, do you think we could pull out and generalize the implementation of beam search in `event2mind` so that it works for `simple_seq2seq` as well?

brendan-ai2 commented:
I would be thrilled to see that. :)

epwalsh commented:
👌 👌  looks do-able. Where would be a good place to put the beam search module?

brendan-ai2 commented:
Perhaps just `allennlp/modules/` for now. I'd have to defer to one of the more seasoned members of the team for the final verdict.

matt-gardner commented:
Beam search isn't a `module` - a pytorch `Module` has parameters that we want to train, and beam search just takes a fixed set of parameters and does search.  I'd default to putting this in `nn/beam_search.py`, unless someone has a better idea.

brendan-ai2 commented:
Thanks for the correction, @matt-gardner!

matt-gardner commented:
Oh, also, @epwalsh, we typically use `scripts/train_fixtures.py` to create those fixture files.  When a model gets updated, feel free to use that to update the model that was changed.

epwalsh commented:
Ahh I see, thanks @matt-gardner!

epwalsh commented:
Hey @matt-gardner and @brendan-ai2 , I think everything is in good shape now, but let me know if you have any other comments. This is what's changed:
- pulled out and generalized the beam search implementation from event2mind, and made a couple small improvements such as halting the decoding early when all predictions are `end_index`.
- updated the unit tests for event2mind
- added unit tests for the new beam search class (`nn.beam_search.BeamSearch`)
- refactored `simple_seq2seq` and added an `attention` module (instead of the legacy attention) and the new beam search for decoding
- updated unit tests for `simple_seq2seq` and retrained the fixture

matt-gardner commented:
@brendan-ai2, can you take this one?  I'm happy to answer any particular questions if you have any.

brendan-ai2 commented:
Of course!

brendan-ai2 commented:
Sorry for the delay @epwalsh and thanks for the ping. I was a bit swamped the last few weeks. Reviewing now.

epwalsh commented:
No worries! Thanks for taking a look

brendan-ai2 commented on allennlp/nn/beam_search.py:
```diff
@@ -0,0 +1,262 @@
+        # within the main loop we are going from the `beam_size` elements of the
+        # beam to `beam_size`^2 candidates from which we will select the top
+        # `beam_size` elements for the next iteration.
+        start_class_log_probabilities, state = first_step(start_predictions, start_state)
```
It doesn't seem like an explicit `first_step` is being passed in anymore, so I'm a bit confused as to how the event2mind model would continue to work. Could you clarify? Sorry if I'm missing something obvious here!

brendan-ai2 replied:
Sorry, got confused here. This is definitely working. :) Is there actually a use case for having `first_step` separate from `step` presently?

epwalsh commented on allennlp/models/encoder_decoders/simple_seq2seq.py:
```diff
@@ -27,221 +25,396 @@ class SimpleSeq2Seq(Model):
+            for each source sentence in the batch.
+        """
+        output_projections, state = self._prepare_output_projections(last_predictions, state)
+        # shape: (group_size, num_classes)
```
No worries, should be any easy fix! I'll make that change in beam search as well

epwalsh commented on allennlp/tests/models/encoder_decoders/simple_seq2seq_test.py:
```diff
@@ -44,6 +38,7 @@ def test_loss_is_computed_correctly(self):
 
     def test_decode_runs_correctly(self):
         training_tensors = self.dataset.as_tensor_dict()
+        del training_tensors["target_tokens"]
```
Oh, this won't be necessary after I address your other comment about the training vs test time logic

epwalsh commented on allennlp/tests/nn/beam_search_test.py:
```diff
@@ -0,0 +1,127 @@
+        self._check_results(batch_size=1)
+
+    def test_greedy_search(self):
+        beam_search = BeamSearch(self.end_index, beam_size=1)
```
Actually, I think this is already covered in the tests

brendan-ai2 replied:
Oh, good point! Sorry about that. For some reason I thought `max_steps` was lower.

epwalsh commented:
Thanks for taking a look @brendan-ai2! Based on your comments, the biggest thing I have left to do is make sure the new model is backwards compatible. That shouldn't be too difficult. I'll let you know when I push that up!

epwalsh commented on allennlp/models/encoder_decoders/simple_seq2seq.py:
```diff
@@ -27,221 +25,396 @@ class SimpleSeq2Seq(Model):
+        state = self._init_encoded_state(source_tokens)
+
+        if target_tokens:
+            return self._forward_train(target_tokens, state)
```
`self._forward_train` has been fixed now

brendan-ai2 replied:
👍

epwalsh commented on allennlp/tests/models/event2mind_test.py:
```diff
@@ -55,22 +61,26 @@ def test_beam_search_orders_results(self):
+        start_predictions = final_encoder_output.new_full(
+                (batch_size,), fill_value=model._start_index, dtype=torch.long)
+        start_state = {"decoder_hidden": final_encoder_output}
+
```
It's been retrained now

brendan-ai2 replied:
Thank you. The validation during training uses beam search. The stopping condition was actually based on a metric computed from a beam, so there was some chance running this would help catch breakage. (Incidentally, we're planning on automating this sort of thing: https://github.com/allenai/allennlp/issues/1813)

epwalsh commented:
Hey @brendan-ai2, I think I've addressed all your comments now. Here's a summary of the latest changes:
- Made attention in `simple_seq2seq` optional, and added support back for legacy attention.
- Changed all other defaults in `simple_seq2seq` back to what they originally were, with the exception of the beam search, so that `simple_seq2seq` is backwards compatible with the original API. I also added tests for this, i.e. I added a `test_model_can_save_train_and_load` for a model with legacy attention and no attention at all.
- Fixed the forward pass in `simple_seq2seq` so that `scheduled_sampling_ratio` is only used during training.
- Retrained the `event2mind` fixture just to make sure everything works.

brendan-ai2 commented on allennlp/models/encoder_decoders/simple_seq2seq.py:
```diff
@@ -27,221 +27,415 @@ class SimpleSeq2Seq(Model):
+        if target_tokens:
+            return self._forward_loop(target_tokens, state)
+
+        return self._forward_beam_search(state)
```
I think running beam search should be guarded behind a flag. Potentially just run it when `beam_size` is defined. The rationale is again backwards compatibility as the shape of the tensors in the returned dictionary is not the same with beam search.

brendan-ai2 commented on allennlp/tests/nn/beam_search_test.py:
```diff
@@ -0,0 +1,171 @@
+        assert list(log_probs.size()) == [batch_size, beam_size]
+        np.testing.assert_allclose(log_probs[0].numpy(), expected_log_probs)
+
+        # check finished state.
```
If you're only testing this once, I'd pull the logic out into the test itself. I try to keep test helpers dead simple to avoid surprises and improve readability.

brendan-ai2 commented on allennlp/tests/predictors/simple_seq2seq_test.py:
```diff
@@ -19,4 +19,7 @@ def test_uses_named_inputs(self):
         assert isinstance(predicted_tokens, list)
-        assert all(isinstance(x, str) for x in predicted_tokens)
+        assert len(predicted_tokens) == predictor._model._beam_search.beam_size
+        for pred_tokens in predicted_tokens:
```
Mmm, might be slightly noisy, but not unreasonable.

brendan-ai2 commented on allennlp/models/encoder_decoders/simple_seq2seq.py:
```diff
@@ -27,221 +25,396 @@ class SimpleSeq2Seq(Model):
+        The attention function to apply over inputs.
+    source_namespace : ``str``, optional (default = 'source_tokens')
+    target_namespace : ``str``, optional (default = 'target_tokens')
+    target_embedding_dim : ``int``, optional (default = 30)
```
I really do appreciate the spirit! Keeping things compatible just inherently limits polish sometimes.

epwalsh commented:
Thanks again for the review @brendan-ai2! I think I've addressed all your comments. The biggest change is that the model now defaults to no beam search. I also left a TODO comment for running beam search during validation, like event2mind, but like you said, it doesn't make sense to do that now since it doesn't support custom metrics yet.
